source/common/filter/config_discovery_impl.cc

```diff
@@ -269,5 +260,85 @@ ProtobufTypes::MessagePtr HttpFilterConfigProviderManagerImpl::getDefaultConfig(
   return message;
 }
 
+std::unique_ptr<DynamicFilterConfigProviderImpl<Http::FilterFactoryCb>>
+HttpFilterConfigProviderManagerImpl::createFilterConfigProviderImpl(
+    FilterConfigSubscriptionSharedPtr& subscription,
+    const absl::flat_hash_set<std::string>& require_type_urls,
+    Server::Configuration::FactoryContext& factory_context,
+    ProtobufTypes::MessagePtr&& default_config, bool last_filter_in_filter_chain,
+    const std::string& filter_chain_type, absl::string_view stat_prefix) {
+  return std::make_unique<HttpDynamicFilterConfigProviderImpl>(
+      subscription, require_type_urls, factory_context, std::move(default_config),
+      last_filter_in_filter_chain, filter_chain_type, stat_prefix);
+}
+
+// TCP listener filter
+absl::string_view TcpListenerFilterConfigProviderManagerImpl::statPrefix() const {
+  return TcpListenerStatPrefix;
+}
+
+ProtobufTypes::MessagePtr TcpListenerFilterConfigProviderManagerImpl::getDefaultConfig(
+    const ProtobufWkt::Any& proto_config, const std::string& filter_config_name,
+    Server::Configuration::ListenerFactoryContext& factory_context, bool, const std::string&,
+    const absl::flat_hash_set<std::string>& require_type_urls) const {
+  auto* default_factory =
+      Config::Utility::getFactoryByType<Server::Configuration::NamedListenerFilterConfigFactory>(
+          proto_config);
+  if (default_factory == nullptr) {
+    throw EnvoyException(fmt::format("Error: cannot find filter factory {} for default filter "
+                                     "configuration with type URL {}.",
+                                     filter_config_name, proto_config.type_url()));
+  }
+  validateTypeUrlHelper(Config::Utility::getFactoryType(proto_config), require_type_urls);
+  return Config::Utility::translateAnyToFactoryConfig(
+      proto_config, factory_context.messageValidationVisitor(), *default_factory);
+}
+
+std::unique_ptr<DynamicFilterConfigProviderImpl<Network::ListenerFilterFactoryCb>>
+TcpListenerFilterConfigProviderManagerImpl::createFilterConfigProviderImpl(
+    FilterConfigSubscriptionSharedPtr& subscription,
+    const absl::flat_hash_set<std::string>& require_type_urls,
+    Server::Configuration::ListenerFactoryContext& factory_context,
```

Comment:
The passed in factory_context reference can be used in `factory->createListenerFilterFactoryFromProto(message, nullptr, factory_context_)`.
When the latter is called, it is unknown whether the factory_context at this line is available. 

More specifically when the created TcpListenerDynamicFilterConfigProviderImpl outlives the ListenerFactoryContext (which is owned by a ListenerImpl),
there is crash risk.

Can you clarify the life time of TcpListenerDynamicFilterConfigProviderImpl and ListenerImpl?

Reply:
Thanks for the comments.

factory_context which is from listener_factory_context_ in class ListenerImpl https://github.com/envoyproxy/envoy/blob/e4eaf1b970876d269f7e15dc049f90c79742e298/source/server/listener_impl.h#L452, has lifetime of ListenerImpl object.  

TcpListenerDynamicFilterConfigProviderImpl creation is planned to be called from here (BTW, the below code path is planned, not fully implemented yet):

https://github.com/envoyproxy/envoy/blob/e4eaf1b970876d269f7e15dc049f90c79742e298/source/server/listener_impl.cc#L617

in below code path:

listener_filter_factories_ = parent_.factory_.createListenerFilterFactoryList()---> createListenerFilterFactoryList()-->. createDynamicFilterConfigProvider()-->createFilterConfigProviderImpl().

 listener_filter_factories_ is created in ListenerImpl c'tor. The lifetime is slightly shorter than listener_factory_context_.

Reply:
Thank you for the detailed explanation! Generally speaking if ConfigProvider is not shard across different ListenerImpl it is good.